pkg/runtime/k0s/init.go

```diff
@@ -36,46 +42,97 @@ func (k *Runtime) init() error {
 	return nil
 }
 
-// MergeConfigOnMaster0 convert the cluster file spec to kubectl.yaml (k0sctl config file) to lead cluster run
-func (k *Runtime) MergeConfigOnMaster0() error {
-	if err := k.K0sConfig.ConvertTok0sConfig(k.cluster); err != nil {
-		return fmt.Errorf("failed to convert to k0s config from clusterfile: %v", err)
+// GenerateConfigOnMaster0 generate the k0s.yaml to /etc/k0s/k0s.yaml and lead the controller node install
+func (k *Runtime) GenerateConfigOnMaster0() error {
+	if err := k.generateK0sConfig(); err != nil {
+		return fmt.Errorf("failed to generate config: %v", err)
+	}
+	k.modifyConfigRepo()
+	if err := k.marshalToFile(); err != nil {
+		return fmt.Errorf("failed to write k0s config: %v", err)
+	}
+	return nil
+}
+
+// GenerateCert generate the containerd CA for registry TLS and k0s join token for 100 years.
+func (k *Runtime) GenerateCert() error {
+	if err := k.generateK0sToken(); err != nil {
+		return err
+	}
+
+	if err := k.GenerateRegistryCert(); err != nil {
+		return err
 	}
-	ssh, err := k.getHostSSHClient(k.cluster.GetMaster0IP())
+	return k.SendRegistryCert(k.cluster.GetMasterIPList()[:1])
+}
+
+func (k *Runtime) generateK0sConfig() error {
+	master0IP := k.cluster.GetMaster0IP()
+	ssh, err := k.getHostSSHClient(master0IP)
 	if err != nil {
-		return fmt.Errorf("failed to get ssh client: %v", err)
+		return err
 	}
-	output, err := ssh.Cmd(k.cluster.GetMaster0IP(), GetK0sVersionCMD)
+	createCMD := "mkdir -p /etc/k0s && k0s config create"
+	bytes, err := ssh.Cmd(master0IP, createCMD)
 	if err != nil {
 		return err
 	}
+	return yaml2.Unmarshal(bytes, &k.k0sConfig)
+}
 
-	k.K0sConfig.DefineConfigFork0s(string(output), k.RegConfig.Domain, k.RegConfig.Port, k.cluster.Name)
+func (k *Runtime) modifyConfigRepo() {
+	k.k0sConfig.Spec.Images.Repository = k.RegConfig.Domain + ":" + k.RegConfig.Port
+}
 
-	//write k0sctl.yaml to master0 rootfs
-	if err := k.K0sConfig.WriteConfigToMaster0(k.getRootfs()); err != nil {
+func (k *Runtime) marshalToFile() error {
```

Comment:
I am afraid that this is not a good design I wish. From my opinion, it is quite a general function to marshal some content to a file. But you designed this functionality to be sub function of struct `Runtime`. Maybe not a best choice.

You could define a general function like `func MarshalTo(xxx interface, file os.filepath) error`, then it is quite general and make the code coherent.